crates/holochain/src/core/state/metadata/keys.rs

```diff
@@ -297,45 +263,95 @@ impl IntoIterator for &LinkMetaKey<'_> {
     type Item = u8;
     type IntoIter = std::vec::IntoIter<Self::Item>;
     fn into_iter(self) -> Self::IntoIter {
-        let b: BytesKey = self.into();
-        b.0.into_iter()
+        (&self).into_iter()
     }
 }
 
-impl IntoIterator for LinkMetaKey<'_> {
-    type Item = u8;
-    type IntoIter = std::vec::IntoIter<Self::Item>;
-    fn into_iter(self) -> Self::IntoIter {
-        (&self).into_iter()
+impl<T: PrefixType> From<&LinkMetaKey<'_>> for PrefixBytesKey<T> {
+    fn from(k: &LinkMetaKey) -> Self {
+        PrefixBytesKey::new(k)
     }
 }
 
-impl IntoIterator for MiscMetaKey {
-    type Item = u8;
-    type IntoIter = std::vec::IntoIter<Self::Item>;
-    fn into_iter(self) -> Self::IntoIter {
-        (&self).into_iter()
+impl<T: PrefixType> From<LinkMetaKey<'_>> for PrefixBytesKey<T> {
+    fn from(k: LinkMetaKey) -> Self {
+        (&k).into()
     }
 }
 
-impl<T: PrefixType> From<MiscMetaKey> for PrefixBytesKey<T> {
-    fn from(k: MiscMetaKey) -> Self {
-        PrefixBytesKey::new(k)
+impl From<&Header> for ChainItemKey {
+    fn from(h: &Header) -> Self {
+        ChainItemKey::Full(
+            h.author().clone(),
+            h.header_seq(),
+            HeaderHash::with_data_sync(h),
+        )
     }
 }
 
-impl<T: PrefixType> From<&LinkMetaKey<'_>> for PrefixBytesKey<T> {
-    fn from(k: &LinkMetaKey) -> Self {
-        PrefixBytesKey::new(k)
+impl From<ChainItemKey> for HeaderHash {
+    fn from(c: ChainItemKey) -> Self {
+        match c {
+            ChainItemKey::Full(_, _, h) => h,
+            _ => unreachable!("Tried to get header hash from a partial key: {:?}", c),
+        }
     }
 }
 
-impl<T: PrefixType> From<LinkMetaKey<'_>> for PrefixBytesKey<T> {
-    fn from(k: LinkMetaKey) -> Self {
-        (&k).into()
+impl From<&ChainItemKey> for BytesKey {
+    fn from(key: &ChainItemKey) -> Self {
+        use byteorder::{NativeEndian, WriteBytesExt};
+        match key {
+            ChainItemKey::Agent(a) => a.as_ref().into(),
+            ChainItemKey::AgentSequence(a, s) => {
+                // Get the agent key
+                let mut buf = a.clone().into_inner();
+                let mut num = Vec::with_capacity(4);
+
+                // Get the header seq
+                num.write_u32::<NativeEndian>(*s).unwrap();
+                buf.extend(num);
+                buf.into()
+            }
+            ChainItemKey::Full(a, s, h) => {
+                // Get the agent key
+                let mut buf = a.clone().into_inner();
+                let mut num = Vec::with_capacity(4);
+
+                // Get the header seq
+                num.write_u32::<NativeEndian>(*s).unwrap();
+                buf.extend(num);
+
+                // Get the header hash
+                buf.extend(h.clone().into_inner());
+                buf.into()
+            }
+        }
+    }
+}
+
+// TODO: This is way to fragile there must be a better way
+// get from the k bytes to the chain item key
```

Comment:
This might need to be better

Reply:
Yeah can you just use some consts, so when we change the consts this changes too?